docs: added notes about HTML5 video elements to working-with-video (#16102)
gatsbyjs/gatsby: docs/docs/working-with-video.md

  <video controls>
    <source src={DogMp4} type="video/mp4" />
    <source src={DogOgg} type="video/ogg" /> // highlight-line
  </video>
)

```

Even though there are two `<source>` elements, only one video will be displayed, first `mp4` if it is supported, then `.ogg`.

**Note**: this requires importing a video in the format of the type specified, i.e. adding a `<source>` element with `type=video/ogg` would also need a file import with a format of `.ogg`. Alternatively, you can specify a URL to where a video is remotely hosted as the `src` instead of importing a local file.

[See an example repository using `<video>` elements](https://github.com/gatsbyjs/gatsby/blob/master/examples/using-video/)

### Accessibility with custom video players

One advantage of integrating a custom component with your own hosted video is it can give you more control over the video player, including its accessibility. Elements of accessible video and audio include:

- captions: a text version of the audio, syncronized with the video
- transcript (or subtitles): a text version of the audio and visual content, like captions but also including descriptions of key visual elements in the video

> - audio description: an audio version of visual information not conveyed in dialouge

Reviewer: ```suggestion
- audio description: an audio version of visual information not conveyed in dialogue
```

- accessible controls: buttons to operate the video that can be operated without a mouse, are labeled, and work across environments and browsers

> Though captions, transcripts, and audio descriptions primarily aim to assist those with greater difficulty seeing or hearing, they benefit many other users who prefer reading to listening.

Reviewer: ```suggestion
Though captions, transcripts, and audio descriptions primarily aim to assist those with greater difficulty seeing or hearing, they benefit many other users who prefer reading to listening. Captions can also help people watching videos when they can't have the sound turned on for whatever reason.
```

HTML5 provides support for these types of assistive content through the `<track>` element. The track element is nested under a `<video`> element as an empty tag. An example usage of the `<track>` element with a video looks like this:

```javascript:title=src/pages/index.js
import React from "react"
import DogMp4 from "../assets/dog.mp4"
import Captions from "file-loader!../assets/captions.vtt" // highlight-line

export default () => (
  <video controls>
    <source src={DogMp4} type="video/mp4" />
    // highlight-start
    <track kind="captions" srcLang="en" src={Captions} />
    // highlight-end
  </video>
)
```

The kind attribute can be of a variety of different types including `captions`, `subtitles`, and `descriptions`, among others. The `srcLang` defines English as the language used in the captions in the example, and the captions file imported is used as the source. You can read about the specific attributes of a [`<track>` on MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/track).

> **Note**: the filepath to import the captions includes the "file-path!" prefix which helps webpack import the `.vtt` file.

Reviewer: ```suggestion
**Note**: the filepath to import the captions in the above code snippet includes the `file-loader!` prefix, which helps webpack import the `.vtt` caption file.
```

HTML5 provides the `<video>` media element for working with videos. Inside the `<video>` element, you can provide multiple `<source>` elements that serve as different file formats the video player can use, with each browser electing to use a format it supports.

If you have a video called `dog.mp4` in your site under `src/assets/dog.mp4`, you can [include that video in your page with webpack](/docs/importing-assets-into-files) like you would other assets. Then reference it in a `<source>` element, which is wrapped by a `<video>` element:

<!-- prettier-ignore -->
```javascript:title=src/pages/index.js
import React from "react"
import DogVideo from "../assets/dog.mp4"

export default () => (
  <video controls style={{ width: "100%" }}>
    <source src={DogVideo} type="video/mp4" /> // highlight-line
  </video>
)

```

> The `controls` attribute on the `<video>` will provide buttons overlaid on the video to play/pause, adjust volume, and go full screen. Other attributes like `muted` can set audio to silent, or `poster` can display an image when the video isn't playing. Common attributes that you'd want to apply to multiple videos could be extracted into a custom React video component. A full list of `<video>` attributes can be found on [the MDN docs](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/video#Attributes).

Reviewer: I thought I worked on this already but maybe I stashed changes and didn't end up adding them - but we should take the opportunity to mention accessibility in video. This can go in a separate PR - but we should mention custom video player UI, transcripts, closed-captions, and audio description for video. This link is a great resource of the things we could mention: https://www.washington.edu/accessibility/videos/

There are also some Gatsby-specific things for audio in this post by Lindsey Kopacz: https://www.a11ywithlindsey.com/blog/a11y-lessons-learned-starting-podcast

I have a few ideas:
- Rename this page to include Working with Video and Audio, and add a few more audio details
- Mention accessibility on this video page and link to a more general page, like [Importing Media Content](https://www.gatsbyjs.org/docs/importing-media-content/)

I'd love to hear your thoughts on this too, @amberleyromo!

Author: Would it make sense to just have a separate page for Working with Audio? I guess some things for accessibility overlap but maybe a lot of that could go on a more general page like Importing Media Content.

I'll add some more info about video in this PR though so it doesn't just sit around waiting in an issue again.

> Adding more source tags for additional formats will allow the browser to find a source type that it supports, if there are no matching source types the video will fail to load. You can see what formats are supported in different browsers on [MDN's docs about supported media formats](https://developer.mozilla.org/en-US/docs/Web/HTML/Supported_media_formats#Browser_compatibility).

Reviewer: ```suggestion
### Supporting multiple browsers and formats

Adding more source tags for additional formats will allow the browser to find a source type that it supports, if there are no matching source types the video will fail to load. You can see what formats are supported in different browsers on [MDN's docs about supported media formats](https://developer.mozilla.org/en-US/docs/Web/HTML/Supported_media_formats#Browser_compatibility).
```

<!-- prettier-ignore -->
```javascript:title=src/pages/index.js
import React from "react"
import DogMp4 from "../assets/dog.mp4"
import DogOgg from "../assets/dog.ogg" // highlight-line

export default () => (
  <video controls style={{ width: "100%" }}>
    <source src={DogMp4} type="video/mp4" />
    <source src={DogOgg} type="video/ogg" /> // highlight-line
  </video>
)

```

Even though there are two `<source>` elements, only one video will be displayed, first `mp4` if it is supported, then `.ogg`.

> **Note**: this requires adding the link to a video in the format of the type specified, i.e. adding `type=ogg` would also need an associated file at `src/assets/dog.ogg` or a link to where that video is hosted elsewhere.

Reviewer: ```suggestion
**Note**: this requires importing a video in the format of the type specified, i.e. adding a `<source>` element with `type=video/ogg` would also need a file import with a format of `.ogg`. Alternatively, you can specify a URL to where a video is remotely hosted as the `src` instead of importing a local file.
```
This was a little confusing to read...let me know what you think!
